### PR TITLE
add RandomCenteredCrop augmentation that depends on the input parameters

### PR DESCRIPTION
**Pull request by LinaShiryaeva**

### CODE REVIEW

File: albumentations/augmentations/functional.py [hunk 1]
```diff
@@ -220,6 +220,17 @@ def random_crop(img, crop_height, crop_width, h_start, w_start):
     img = img[y1:y2, x1:x2]
     return img
 
+def simple_crop_coordinates(img, x_min, x_max, y_min, y_max):
```
Comment (albu) [inline albumentations/augmentations/functional.py]: crop_image_by_coordinates

consider adding same method for bounding boxes

File: albumentations/augmentations/transforms.py [hunk 2]
```diff
@@ -459,6 +459,50 @@ def get_params(self):
     def apply_to_bbox(self, bbox, **params):
         return F.bbox_random_crop(bbox, self.height, self.width, **params)
 
+class RandomCenteredBBoxCrop(DualTransform):
```
Comment (albu) [inline albumentations/augmentations/transforms.py]: RandomCropNearBbox

File: albumentations/augmentations/transforms.py [hunk 3]
```diff
@@ -459,6 +459,50 @@ def get_params(self):
...
+    #TODO comments
```
Comment (albu) [inline albumentations/augmentations/transforms.py]: resolve this todo

File: tests/test_augmentations.py [hunk 4]
```diff
@@ -83,10 +83,12 @@ def test_image_only_augmentations_with_float_values(augmentation_cls, params, fl
     [CenterCrop, {'height': 10, 'width': 10}],
     [RandomCrop, {'height': 10, 'width': 10}],
     [RandomSizedCrop, {'min_max_height': (4, 8), 'height': 10, 'width': 10}],
+    [RandomCenteredBBoxCrop, {'max_part_shift': 0.15}]
 ])
 def test_dual_augmentations(augmentation_cls, params, image, mask):
     aug = augmentation_cls(p=1, **params)
-    data = aug(image=image, mask=mask)
+    annotations = {'image': image, 'cropped_bbox': [-59, 77, 177, 231]}
```
Comment (albu) [inline tests/test_augmentations.py]: you should not change tests, just add another if existing does not fit your needs.

File: tests/test_augmentations.py [hunk 5]
```diff
@@ -106,10 +108,12 @@ def test_dual_augmentations(augmentation_cls, params, image, mask):
     [CenterCrop, {'height': 10, 'width': 10}],
     [RandomCrop, {'height': 10, 'width': 10}],
     [RandomSizedCrop, {'min_max_height': (4, 8), 'height': 10, 'width': 10}],
+    [RandomCenteredBBoxCrop, {'max_part_shift': 0.15}]
 ])
 def test_dual_augmentations_with_float_values(augmentation_cls, params, float_image, mask):
     aug = augmentation_cls(p=1, **params)
-    data = aug(image=float_image, mask=mask)
+    annotations = {'image': float_image, 'cropped_bbox': [-59, 77, 177, 231]}
```
Comment (albu) [inline tests/test_augmentations.py]: you should not change tests, just add another if existing does not fit your needs.

File: tests/test_augmentations.py [hunk 6]
```diff
@@ -179,7 +184,8 @@ def test_augmentations_wont_change_input(augmentation_cls, params, image, mask):
     image_copy = image.copy()
     mask_copy = mask.copy()
     aug = augmentation_cls(p=1, **params)
-    aug(image=image, mask=mask)
+    annotations = {'image': image, 'cropped_bbox': [-59, 77, 177, 231]}
```
Comment (albu) [inline tests/test_augmentations.py]: you should not change tests, just add another if existing does not fit your needs.

File: tests/test_augmentations.py [hunk 7]
```diff
@@ -221,7 +228,8 @@ def test_augmentations_wont_change_input(augmentation_cls, params, image, mask):
 def test_augmentations_wont_change_float_input(augmentation_cls, params, float_image):
     float_image_copy = float_image.copy()
     aug = augmentation_cls(p=1, **params)
-    aug(image=float_image)
+    annotations = {'image': float_image, 'cropped_bbox': [-59, 77, 177, 231]}
```
Comment (albu) [inline tests/test_augmentations.py]: you should not change tests, just add another if existing does not fit your needs.

File: albumentations/augmentations/functional.py (hunk 1 above)
Comment (LinaShiryaeva) [reply albumentations/augmentations/functional.py]: "consider adding same method for bounding boxes" - what does this method should do?

File: albumentations/augmentations/transforms.py (hunk 2 above)
Comment (LinaShiryaeva) [reply albumentations/augmentations/transforms.py]: Is there is I need only to change the class name?

File: albumentations/augmentations/functional.py (hunk 1 above)
Comment (albu) [reply albumentations/augmentations/functional.py]: crop bounding boxes by same "crop_box". new feature should work for all existing targets

File: albumentations/augmentations/transforms.py (hunk 2 above)
Comment (albu) [reply albumentations/augmentations/transforms.py]: yes

File: albumentations/augmentations/transforms.py [hunk 8]
```diff
@@ -459,6 +459,54 @@ def get_params(self):
...
+        return {'x_min': x_min,
+                'x_max': x_max,
+                'y_min': y_min,
+                'y_max': y_max
+                }
+
+    def apply_to_bbox(self, bbox, x_min = 0, x_max = 0, y_min = 0, y_max = 0, **params):
+        h_start = y_min
+        w_start = x_min
+        return F.bbox_random_crop(bbox, y_max - y_min, x_max - x_min, h_start, w_start, **params)
```
Comment (albu) [inline albumentations/augmentations/transforms.py]: why random? it's bbox_crop

File: albumentations/augmentations/transforms.py [hunk 9]
```diff
@@ -459,6 +459,54 @@ def get_params(self):
...
+
+    @property
+    def targets_as_params(self):
+        return ['cropped_bbox']
```
Comment (albu) [inline albumentations/augmentations/transforms.py]: maybe cropping_bbox is better?

File: tests/test_augmentations.py [hunk 10]
```diff
@@ -317,3 +317,13 @@ def test_augmentations_wont_change_shape_rgb(augmentation_cls, params, image, ma
     result = aug(image=image_3ch, mask=mask_3ch)
     assert np.array_equal(image_3ch.shape, result['image'].shape)
     assert np.array_equal(mask_3ch.shape, result['mask'].shape)
+
+@pytest.mark.parametrize(['augmentation_cls', 'params'], [
+    [RandomCropNearBBox, {'max_part_shift': 0.15}]
+    ])
+def test_image_only_random_crop_augmentation(augmentation_cls, params, image, mask):
```
Comment (albu) [inline tests/test_augmentations.py]: change name to test_image_only_crop_around_bbox_augmentation

**Comment by albu** [comment]
lets squash everything and push --force into this branch

**Pull request merged**